src/ledger/test/LedgerTxnTests.cpp

```diff
@@ -2804,6 +2819,178 @@ TEST_CASE("Erase performance benchmark", "[!hide][erasebench]")
+        {
+            REQUIRE(lkMeter->canLoad(LedgerEntryKey(e), xdr::xdr_size(e)));
+            expectedSuccessKeys.emplace(LedgerEntryKey(e));
+        }
+
+        // Insert entry into set with correct lastModifiedLedgerSeq value so
+        // we can check prefetch results later
+        e.lastModifiedLedgerSeq = 1;
+        entrySet.emplace(e);
+        entryMap.emplace(LedgerEntryKey(e), e);
+    }
+    if (cfg.EXPERIMENTAL_BUCKETLIST_DB)
+    {
+        std::vector<LedgerEntry> ledgerVect{entrySet.begin(), entrySet.end()};
+        app->getBucketManager().addBatch(*app, 2, 20, {}, ledgerVect, {});
+    }
+    ltx.commit();
+
+    SECTION("prefetch normally")
+    {
+        LedgerTxn ltx2(root);
+        UnorderedSet<LedgerKey> smallSet;
+        LedgerKeySet expectedSuccessKeys2;
+        LedgerKeySet expectedFailedKeys2;
+
+        for (auto const& k : keysToPrefetch)
+        {
+            auto entry = entryMap.at(k);
+            UnorderedSet<LedgerKey> ks{k};
+            stellar::uniform_int_distribution<uint32_t> dist(1, 10);
+            int32_t offset = 5 - dist(stellar::gRandomEngine);
+            // store the initial quota for the transaction.
+            auto quota = xdr::xdr_size(entry) + offset;
+            lkMeter2->addTxn(quota, ks);
+            if (offset < 0)
+            {
+                expectedFailedKeys2.emplace(k);
+                REQUIRE(!lkMeter2->canLoad(k, xdr::xdr_size(entry)));
+            }
+            else
+            {
+                expectedSuccessKeys2.emplace(k);
+                REQUIRE(lkMeter2->canLoad(k, xdr::xdr_size(entry)));
+            }
+            smallSet.emplace(k);
+            if (smallSet.size() > (cfg.ENTRY_CACHE_SIZE / 3))
+            {
+                break;
+            }
+        }
+        auto loadSize = root.prefetchSoroban(smallSet, lkMeter2);
+        for (auto const& k : expectedFailedKeys2)
+        {
+            REQUIRE(lkMeter2->isNotLoaded(k));
+        }
+        REQUIRE(loadSize == expectedSuccessKeys2.size());
+
+        // Check that prefetch results are actually correct
+        for (auto const& k : smallSet)
+        {
+            if (expectedFailedKeys2.find(k) != expectedFailedKeys2.end())
+            {
+
+                REQUIRE(lkMeter2->isNotLoaded(k));
+                continue;
+            }
+            REQUIRE(!lkMeter2->isNotLoaded(k));
+            auto txle = ltx2.load(k);
+            REQUIRE(txle);
+            REQUIRE(entrySet.find(txle.current()) != entrySet.end());
+        }
+
+        // 100% hit rate but make it floating point
+        REQUIRE(fabs(ltx2.getPrefetchHitRate() - 1.0f) < 0.0001f);
+        ltx2.commit();
+    }
+    SECTION("prefetch more than ENTRY_CACHE_SIZE entries")
+    {
+        LedgerTxn ltx2(root);
+        auto result = root.prefetchSoroban(keysToPrefetch, lkMeter);
+        REQUIRE(result == expectedSuccessKeys.size());
+        ltx2.commit();
+    }
+}
+
+TEST_CASE("LedgerKeyMeter tests")
```

Comment:
If I understand correctly, it looks like "Soroban Prefetch" tests the following case:
All txs have 1 unique Soroban metered key (as in no TTL entry). Some txs have enough resources, some do not. Check that they are loaded or not loaded properly. No TXs have overlapping keys.

I see you have some additional edge cases covered in the `LedgerKeyMeter tests` which is good, but I'd like to see this edge cases tested more end to end at the ledgertxn level. I think adding the following cases to "Soroban Prefetch" would be helpful (even if some of these are redundant with "LedgerKeyMeter tests":

1. Mix of metered soroban and classic keys
2. Multiple TXs referencing the same key, some of which have enough resource and some do not. I think it would be interesting if there was a TX that read 3 keys and only ran out of budget on the last one. Then maybe another TX that only reads failed key from the previous TX but has enough budget. Just interesting permutations of TXs that are overlapping that have a few keys instead of just one key.
3. Live cached entries are metered
4. Dead cached entries (i.e. null entries) are not metered (this is currently not working)
5. TTL keys are not metered